### ISSUE TITLE
Missing documentation

### ISSUE THREAD

**Issue by FibreFoX** [FibreFoX/javafx-gradle-plugin#8]

**Comment by FibreFoX** [comment]
See #5 #6 #7

**Comment by JustGregory** [comment]
If this is open for commentary, I would also like to add that I would like to see another example of the buildscript, based upon my intended use-case. (This is the third time typing this, as I lost the post from the last shutdown/restart; please bear with me for typing at length.)

I'm creating a meta-project-series, intended to be my own implementation of an "RPG Maker", which contains a game editor, a game engine, and a platform architecture. In this example, the game engine and the platform modules are plain-Java classes because for an RPG, FXML may be desireable but often a less-than-optimal solution; but the editor is intended to be using a *lot* of JavaFX FXML. But the editor is broken up into subprojects, one or two of the subprojects will use FXML features, and be included in the editor core for extensibilitiy, but only the "editor" core project of the subprojects will be meant to be built as a JavaFX-enabled application.

I'm not specifically sure, but I believe your example in the readme was intended for projects which are "all-in-one" build solutions, which compicates things when subprojects are used. For me, with distributing functionality across a number of modules, I am specifically curious upon the placement of which lines from the example in each build gradle file would result in the most optimal solution for my specific case. I use NetBeans IDE and the Gradle plugin for it developed by @kelemen (Attila Kelemen) which is actively developed. Creating new Gradle root projects produce very specific ```build.gradle```, ```common.gradle```, and settings.gradle``` files; then subprojects have very minimal ```build.gradle``` lines themselves.

Now, I know that most if not all things in the parent ```build.gradle``` are also passed to subprojects, but for things which specifically need to be in the subprojects, it's useful to use a ```common.gradle``` that is included in the main build script. I'm tinkering right now with such a project, trying to see what placement or ordering of lines for parent/child project structures would be most optimized (or least non-optimal, if you want to split hairs). If you have a specific idea of what this should be, I'll go with it; otherwise, I'll continue experimenting, rebuilding, editing, and so on.

I feel that one specific README may become quite complicated with describing some usages of this Gradle plugin. To that end, if you could create a Github Pages simple website, you could expand the documentation and varieties of buildscript usage more than just one Readme file can hold. Then, you merely have to note in this readme that "a more complete description of its usage is located in the Github Pages" or some such line.

**Comment by FibreFoX** [comment]
The Readme-file on this project was the minimum-description I could think of. This is not the final state, and I intent to implement most or even all project-examples/-tests like inside the maven-plugin.

With the [javafx-maven-plugin](https://github.com/javafx-maven-plugin/javafx-maven-plugin) you normally implement your application and want it to be created with JFX-magic. All dependencies (external, wich have group-id, artifact-id and mostly a version) are described within `pom.xml`-file. There exists another concept: having a project wich just creates an empty jar-file (or some readme/license-files etc.) but [bundles all needed jars/dependencies](https://github.com/javafx-maven-plugin/javafx-maven-plugin/tree/master/src/it/16-multi-module-project) which still are described within that `pom.xml`.

When having some submodules, you can declare dependencies which are just file-references (which should be created for each module/plugin you are creating:
```
dependencies{
    // when you are having a maven-dependency:
    compile "com.some.group:dependency:1.0"
    // when having some already generated jar-files 
    compile files("your-engine.jar")
    compile files("../anotherModule/build/libs/anotherModule.jar")
    // when you want to use user-overridable jvm-settings
    // see here: https://docs.oracle.com/javase/8/docs/technotes/guides/deploy/jvm_options_api.html
    // examples: https://github.com/javafx-maven-plugin/javafx-maven-plugin/blob/master/src/it/11-access-application-and-jvm-parameters/src/main/java/com/zenjava/test/Main.java
    compile files("${System.properties['java.home']}/../lib/packager.jar")
}
```

To have some clue about the configuration-names, just look at two places:
* http://javafx-maven-plugin.github.io/ (a bit outdated ... my fault ... not that much time currently)
* inside the [gradle-extension](https://github.com/FibreFoX/javafx-gradle-plugin/blob/master/src/main/java/de/dynamicfiles/projects/gradle/plugins/javafx/JavaFXGradlePluginExtension.java) of this gradle-plugin

As this is the mayor use-case of a jar-generation-process (and to make sure to have some java-project at all), the jfxJar-task [relies on the jar-task](https://github.com/FibreFoX/javafx-gradle-plugin/blob/master/src/main/java/de/dynamicfiles/projects/gradle/plugins/javafx/JavaFXGradlePlugin.java#L60) (which generates some jar-file).

How you have to adjust your gradle-buildscript ... I doubt thats on your side.

As my private time is up for today, there should be some better documentation on the next update. The way will might be some github-hosted pages, not decided yet. But I really hope this gradle-plugin works for you somehow.

**Comment by JustGregory** [comment]
Okay, but just for clarity, you are saying as long as I put the example script into the root project's ```build.gradle```, (adjusting for the particulars of my project) then it doesn't matter which subproject has the JavaFX main class and FXML resources, everything automatically compiles without me being required to break up the example into "this part for the parent, that part for the subproject," and so on?

**Comment by FibreFoX** [comment]
Thats right, it does not matter which JAR does contain the javafx-main-class :smile_cat: the generated `-jfx.jar` contains some special file: `generated-jfx.jar:/META-INF/MANIFEST.MF`

That file works like a "boot-loader", because it contains all required jar-files which should go to the classpath. So all you have to put together: all your dependencies and all your modules, let them be wrapped together by this gradle-plugin, and create a slim-sized jar-file. That way you can create the launcher and all modules in seperate packages/subprojects/external-dependencies but then just bundle them all together.

Does this help you for your goal?

**Comment by JustGregory** [comment]
That's getting me there, thank you. If any further issues/questions arise, I will raise them later.

**Comment by cconnoll** [comment]
I don't know if this helps/needs to be said but in relation to https://github.com/FibreFoX/javafx-gradle-plugin/issues/5 and https://github.com/FibreFoX/javafx-gradle-plugin/issues/7, I understand that ````src/main/deploy```` is not necessarily needed however if you do go by the [README](https://github.com/FibreFoX/javafx-gradle-plugin/blob/master/README.md) and use "src/main/deploy" like:

    jfx { 
     .........
    appName = "project"
    deployDir = "src/main/deploy"
     .........
    }

Then (at least for Windows) the images need to be placed in:

    src/main/deploy/package/windows

where src is at the the base/root of the project.

I've have had to add ${appName}.ico and ${appName}-setup-icon.bmp images to that folder to get them to be packaged up.

**Comment by FibreFoX** [comment]
@cconnoll thanks for the hint, to be honest, the structure is made up by the javapackager itself, like described in [the official documentation by oracle](http://docs.oracle.com/javase/8/docs/technotes/guides/deploy/self-contained-packaging.html#BCGICFDB), look at the [sample output](http://docs.oracle.com/javase/8/docs/technotes/guides/deploy/self-contained-packaging.html#BCGHHDGC):
```
Using base JDK at: /Library/Java/JavaVirtualMachines/jdk1.7.0_06.jdk
  Using default package resource [Bundle config file] (add
    package/macosx/Info.plist to the class path to customize)
  Using default package resource [icon] (add package/macosx/DemoApp.icns
      to the class path to customize)
Creating app bundle: /tmp/test/TestPackage/bundles/DemoApp.app
Config files are saved to /var/folders/rd/vg2ywnnx3qj081sc5pn9_
    vqr0000gn/T/build703997
```

Most important parts are `package/macosx/Info.plist` and `package/macosx/DemoApp.icns`

About that folder-thing: you are right, that **the folder `src/main/deploy` is not required**

**Comment by amaral7** [comment]
I have some configuration questions. Maybe looks like a little bit simple, but I haven't find a way:

- I would like to Bundle JRE
- Set the desktop icon

Can you describe how to set those configurations?

**Comment by FibreFoX** [comment]
Hi @amaral7 ,
to have bundled JRE, just don't set runtime :) it's the normal behaviour to bundle the JRE when executing the jfxNative-task (`gradle jfxNative`). To set desktop-icons, what do you mean? Which OS do you target? Windows?

**Comment by amaral7** [comment]
Hi @FibreFoX, Thank you, the bundle JRE worked. My OS targets are Windows and Mac. I want to change the .EXE icon, today it's a Java Gray icon. Here is a picture of my /build/jfx/native folder:

![application icon](https://cloud.githubusercontent.com/assets/1706819/15323587/155773e4-1c19-11e6-96cf-050a66053cf3.png)

I want to change those icons indicated with a red arrow.

**Comment by FibreFoX** [comment]
Oh I see, this is normal javapackager/javafx-maven-plugin behaviour. Just create some folder in your project with some icon inside having the specified `<appName>`.
For windows: `src/main/deploy/windows/` + `Vimeo Video Uploader.ico`
For mac: `src/main/deploy/macosx/` + `Vimeo Video Uploader.icns`

I'm really irritated sometimes, that using some "plugin" makes the developers not looking at the original documentation of oracle :) but hey, thats okay for me.

As I'm currently fixing [some bug inside the javafx-maven-plugin](https://github.com/javafx-maven-plugin/javafx-maven-plugin/issues/205) (which might get migrated to the gradle-plugin afterwards) and developing a new configurator-website for that plugin, and if that all is finished, I'm going to create a configurator-website for the javafx-gradle-plugin too ;)

**Comment by cconnoll** [comment]
I don't know if it helps but this is something that I had trouble with for a uni project. Screenshots of my project structure and the related properties in my build.gradle (as per @FibreFoX's comment) are as follows:

<img width="444" alt="file structure" src="https://cloud.githubusercontent.com/assets/13199498/15324481/f3a971e0-1c89-11e6-8a9b-8eb939cf2a93.png">
<img width="315" alt="build" src="https://cloud.githubusercontent.com/assets/13199498/15324485/f774f77c-1c89-11e6-81d4-7d8561d92586.png">

**Comment by amaral7** [comment]
@FibreFoX Thanks again, I'm sorry for that. Maybe I should look more deep, I really read a lot of documentations from oracle about javafx and icons, but I just found how to change icon using the stage (`primaryStage.getIcons().add(`), anyway thanks again.

**Comment by amaral7** [comment]
@cconnoll Thank you, helps a lot. Maybe I would have some issues to set the deployDir.

**Comment by FibreFoX** [comment]
@amaral7 it is NOT required to change deployDir, it is per-default set to `src/main/deploy`, but this depends on how you are configuraing your buildscript, as you can see [in the proguard-example](https://github.com/FibreFoX/javafx-gradle-plugin/blob/master/examples/javafx-proguard-gradle/build.gradle#L34)

**Comment by amaral7** [comment]
@FibreFoX Thanks, it works.

**Comment by drkbkr** [comment]
This seems to have become a general support thread, so I'll leave my question here. Apologies if that's not appropriate. 

JavaFx applications support named and unnamed parameters to the application itself. Is there a way to specify those (I'm particularly interested in named ones) in the jfx block within build.gradle? I didn't see any mention of them in the example build.gradle on the front page here. I know, for instance, that the retired javafx-gradle plugin developed by shemnon supported an 'arguments' property (though I never figured out named parameters). Anything like that here? 

Thanks!

**Comment by FibreFoX** [comment]
Hi @drkbkr and thanks for joining here ;)

do you have something like this in mind? [Oracle-documentation about `<fx:jvmarg>`, `<fx:jvmUserArg>` and `<fx:property>`](https://docs.oracle.com/javase/8/docs/technotes/guides/deploy/self-contained-packaging.html#r17c1-t4).

When you are talking about "normal application arguments" (`<fx:argument>`), maybe you want to have a look at the `launcherArguments`, which is a normal list of `String`-entries.

Does this help you?

**Comment by drkbkr** [comment]
Thanks @FibreFoX . 

I don't think it's `launcherArguments` I'm looking for. 

Essentially what it is is a way to specify (within build.gradle) arguments to the main() method. Generally those get handed to the `Application`'s `launch` method which then makes them available as an instance of `Parameters` through the `getParameters()` method. 

For instance, using shemnon's retired javafx plugin you could write something like 

`javafx {
    ...
    arguments = ['argOne', 'argTwo']
    ....
}`

in build.gradle and whatever was specified in that `arguments` field would be passed to the main() method. `launcherArguments` doesn't seem to do that.

Hopefully this is clear. I'm not that familiar with JavaFx apps. I'm just the guy migrating ours from ant to gradle.

Thanks again!

**Comment by FibreFoX** [comment]
Hi @drkbkr, have you tried to replace `arguments = []` with `launcherArguments = []` ? We are talking about the same thing here :smile_cat: just checked that with a dummy-project (which might get uploaded as some example with the next update).

**Comment by drkbkr** [comment]
Maybe I'm running the app the wrong way? I'm just running `./gradlew jfxRun` and the launcher arguments don't seem to make it in.

**Comment by FibreFoX** [comment]
@drkbkr yes, you are running it wrong, because the `jfxRun`-task executes the generated JAR-file, not the (potentially generated) native launcher. That launcherArguments/arguments-parameter is for the native launcher, not the javafx-jar-executable itself.

**Comment by drkbkr** [comment]
Yes, just ran jfxNative and then the resulting executable and there they were. Thanks for the help (and patience)!

**Comment by BrianRichJohn** [comment]
Hi FibreFoX

Love what is brewing here, Joind Gihub especially for this project. when I run the jfxJar Task. it does not seem to include the FXML in the Jar. I have looked in your example back and forth and I have to admit I'm a bit stumped here. Please can you assist if you dont mind

**Comment by FibreFoX** [comment]
Hi to all,

this might get out of hand, as this was not really intended as some support-channel/-thread. I will lock this issue right here (appologies to all about that).

When you have new questions, and think this is a bug, please create a separate/new issue. When you require some assistence, don't hesitate to send me [some email to me](mailto:fibrefox@dynamicfiles.de). I will do my very best to help you with your problems.

I'm actually preparing some example-projects ([on a separate branch](https://github.com/FibreFoX/javafx-gradle-plugin/tree/more_examples/examples) which will get merged after being finished). When these projects are made (they will include some small readme-file), this issue will get closed.

@BrianRichJohn please send me some e-mail, I will respond do you.
